doc/dev/static_type_checking.md

```diff
@@ -0,0 +1,1452 @@
+3) Type hints act as documentation in code and inform the reader of the expected types. This benefits us - the developers of the Python SDK!
+
+Since Python 3, type annotations were introduced and are the **preferred** method of adding type hints to your code.
+Annotations provide a cleaner syntax and let you keep the type information closer, or inline with the code.
+
+```python
+from typing import Any, Union
+
+def download_blob_from_url(
+        blob_url: str,
+        output: str,
+        credential: Union[AzureNamedKeyCredential, TokenCredential],
+        **kwargs: Any
+) -> None:
+    ...
+```
+
+> Note: Do not use comment style type hints (`# type: str`). Some of our libraries use type comments due to legacy code supporting Python 2, but these will be updated to annotation style.
+
+A fully annotated signature includes type annotations for all parameters and the return type. The type of a parameter
+should follow the `:` syntax and a default argument can be supplied like in the `credential` parameter above. A return
+type follows the function def with an arrow `->`, its type, and then closes with `:`.
+
+It is also possible to add type annotations to variables. The syntax follows the same as
+function arguments:
+
+```python
+length: int = 42
+width: int = 8
+area: int = length * width
+```
+
+When typing your library, note that almost anything can be used as a type - basic data types like `str`, `int`, `bool`,
+types from the `typing` or `typing_extensions` modules, user-defined types, abstract base classes, and more.
+See [Types usable in annotations](#types-usable-in-annotations) for more information.
+
+## Typing a client library
+
+When starting to add type hints to a client library, there are a few things to consider. First, Python has a "gradual
+type system". What this means is that typing is not an all-or-nothing task - you can gradually add types into your code
+and any code without type hints will just be ignored by the type checker. Typing is optional and this is a good
+thing! Pushing to achieve full coverage of annotated code can lead to low signal-to-noise and sometimes is not practical 
+given the expressiveness of Python as a language. So, in practice, what should you aim to type?
+
+1) Add type hints to publicly exposed APIs in the client library. Type hints get shipped with our client libraries (both whl and sdist)
+   and provide benefit to our users. To ensure that type hints do get shipped in your library and can be checked by a static
+   type checker, follow the steps per [PEP 561](https://mypy.readthedocs.io/en/stable/installed_packages.html#creating-pep-561-compatible-packages) below:
+
+    - add an empty `py.typed` file to your package directory. E.g. `.../sdk/azure-core/azure/core/py.typed`
+    - include `py.typed` under `package_data` in your setup.py (`package_data={"azure-core": ["py.typed"]}`) ([example](https://github.com/Azure/azure-sdk-for-python/blob/main/sdk/core/azure-core/setup.py))
```

Comment:
question @kristapratico: it looks like the setup.py file in azure-core has the following line:
`package_data={"pytyped": ["py.typed"]}`.
Here, you say, "azure-core" instead of "pytyped". Is there a difference?